### PR TITLE
Synchronize the token type selector with the permissions in the request

### PR DESCRIPTION
**Pull request by simotae14**
As an admin user managing API tokens settings, I want the token type selector to provide me the possibility of defining a read-only or full access token, or in alternative a custom set of permissions.

👨🏽‍💻 Details
Read-only and full access token types are not simply a way to fastly select a specific set of permissions, they’re identifiers of a set of permissions.

An example to clarify:
 - if a user has a full access token, he will have the permissions also selected on, for example, API created after the creation of a token
 - if a user has a custom token with all the permissions selected on all the existing APIs, she will have the permissions only on the selected APIs and not, for example, on APIs created after the creation of a token

☑️ Acceptance criteria

 - As an admin user managing the settings of an API token
   - when I select Read-only as a Token type
     - then I want all the permissions “frozen” / not editable 
     - and then I want all the find and findOne permissions to be checked
   - when I select Full access as a Token type
     - then I want all the permissions “frozen” / not editable
     - and then I want all the permissions to be checked
   - when I select custom as Token type
     - then I want all the permissions selectable
     - and then I want the permissions checked to correspond with the last interaction of the UX with the permissions 
       - in particular 
         - when I create a token and I directly set the Custom token type
           - then I want no permissions checked
         - when I create a token and I select Read-only token type and then Custom token type
           - then I want the find and findOne permissions checked (inherited from the last interaction on the Token type)
        - when I create a token and I select Full access token type and then Custom token type
          - then I want all the permissions to be checked (inherited from the last interaction on the Token type)

### CODE REVIEW

File: packages/core/admin/admin/src/pages/SettingsPage/pages/ApiTokens/EditView/index.js
Comment (remidej) [inline packages/core/admin/admin/src/pages/SettingsPage/pages/ApiTokens/EditView/index.js:296]: Mutating a render prop inside the render seems weird to me. Is there a way to do it it in an event listener for example?

File: packages/core/admin/admin/src/pages/SettingsPage/pages/ApiTokens/EditView/index.js
Comment (remidej) [inline packages/core/admin/admin/src/pages/SettingsPage/pages/ApiTokens/EditView/index.js:478]: Have you tested what happens if values is nullish? You may need to give a fallback value (`__null_reset_value__`)

**Review by remidej — COMMENTED** [review]
Thanks for the detailed PR notes, makes it easier to review. The code looks good I just left a couple of comments 🙂 

Side note: 500+ lines in a single file doesn't seem very manageable to me, I'd recommend extracting some components in a subfolder

File: packages/core/admin/admin/src/pages/SettingsPage/pages/ApiTokens/EditView/index.js
Comment (simotae14) [reply packages/core/admin/admin/src/pages/SettingsPage/pages/ApiTokens/EditView/index.js:296]: No unfortunately because I need to overwrite the api token type to custom only when we create an api token and when we select for the first time a permission (I need to have access to the form state, values in the example)

File: packages/core/admin/admin/src/pages/SettingsPage/pages/ApiTokens/EditView/index.js
Comment (simotae14) [reply packages/core/admin/admin/src/pages/SettingsPage/pages/ApiTokens/EditView/index.js:478]: the type can be or undefined or one of these values ('full-access', 'custom', 'read-only'), there are not other possible values but if it would be null it works like the undefined one

Comment (simotae14) [reply packages/core/admin/admin/src/pages/SettingsPage/pages/ApiTokens/EditView/index.js:487]: if it is null we show a validation error

File: packages/core/admin/admin/src/pages/SettingsPage/pages/ApiTokens/EditView/index.js
Comment (simotae14) [reply packages/core/admin/admin/src/pages/SettingsPage/pages/ApiTokens/EditView/index.js]: I can set an initial value, I think can be a better solution

**Review by remidej — APPROVED** [review]
Thanks for your responses and the update

**Review by Convly — CHANGES_REQUESTED** [review]
When setting the type to custom by checking some permissions, then trying to go back to read-only or full-access, it changes the checked permissions, but not the token type (it goes back to custom)
peek

File: packages/core/admin/admin/src/pages/SettingsPage/pages/ApiTokens/EditView/index.js
Comment (Convly) [inline packages/core/admin/admin/src/pages/SettingsPage/pages/ApiTokens/EditView/index.js:503]: I don't think you want to keep this :slightly_smiling_face:

**Review by Convly — CHANGES_REQUESTED** [review]

**Comment by Convly** [comment]
Also, I now have this weird behavior where:

1. Create a new token
2. See that the type is `Select`
3. Click on a checkbox, the type is now `Custom`
4. Uncheck the checkbox, the type goes back to `Select`

However, if you do

1. Create a new token
2. See that the type is `Select`
3. Click on a checkbox, the type is now `Custom`
4. Click on the save button
5. Uncheck the checkbox, the type stay `Custom`

Might be something to do with a comparison you're doing using the token's type (the one returned from the API) instead of the type stored in the state? (I'm just fishing here, I don't really know what kind of values you're using)

Here's a small recording of the behavior
custom

**Review by Convly — APPROVED** [review]
LGTM in terms of functionalities, might want to have a new look at the code from @remidej  :eyes:

File: packages/core/admin/admin/src/pages/SettingsPage/pages/ApiTokens/EditView/index.js
Comment (remidej) [inline packages/core/admin/admin/src/pages/SettingsPage/pages/ApiTokens/EditView/index.js:315]: Sorry to bring this up again, but it's the same pattern (mutations in render) as the previous commit, which I'm not sure is recommended. Maybe you could try extracting it in a subcomponent, so that you could do that logic outside the return, or in an effect for example

I'd love to get a second opinion on this topic though, as it may not be a problem, especially since JS found no issues in testing

Comment (simotae14) [reply packages/core/admin/admin/src/pages/SettingsPage/pages/ApiTokens/EditView/index.js:315]: hey Remi I have refactored again the code now using a Formik features called setFieldValue https://formik.org/docs/api/formik#setfieldvalue-field-string-value-any-shouldvalidate-boolean--void and I use this utility only when the Permissions context change and the type is empty

with this solution I have fixed all the bugs.
Related to your suggestion, unfortunately the functions triggered when we change the Permissions are outside the Formik Component because are related to the ApiTokenPermissionsContextProvider, and this component wraps the Formik form and others components

**Review by remidej — APPROVED** [review]
Thanks for finding that setFieldValue API, it's perfect here

**Comment by innerdvations** [comment]
I got errors until I merged in `features/api-token-v2` again, but now it works fine. LGTM! Once CI tests finish we can merge.

**Review by innerdvations — APPROVED** [review]

**Pull request merged**
